### DISCUSSION TITLE
Pruning Citrinet model

### DISCUSSION THREAD

**Discussion by amoghatr** [NVIDIA-NeMo/NeMo#4206] [Q&A]
Hi, I'm trying to reduce the size of my trained citrinet model. I loaded the checkpoint in pytorch and tried to prune it according to the official pytorch tutorial but ran into errors. How to prune it efficiently?

**Comment by titu1994** [comment]
Could you describe what steps you performed to prune a model, and the stack trace of what error you faced ?

We haven't tried to prune Citrinets, though we do somewhat support scripts that perform quantization (though again, they were not tested on Citrinets but instead on QuartzNet)

**Reply by amoghatr** [reply]
```
for state_param in model.state_dict():
    print(state_param)
```
output:
encoder.0.mconv.0.conv.weight
encoder.0.mconv.1.conv.weight
encoder.0.mconv.2.layer.0.weight
encoder.0.mconv.2.layer.0.bias
encoder.0.mconv.2.layer.0.running_mean
encoder.0.mconv.2.layer.0.running_var
encoder.0.mconv.2.layer.0.num_batches_tracked
encoder.0.mconv.3.layer.0.fc.0.weight
encoder.0.mconv.3.layer.0.fc.2.weight
encoder.1.mconv.0.conv.weight
encoder.1.mconv.1.conv.weight
encoder.1.mconv.2.layer.0.weight (continues)

after this, I tried to prune the first two tensors by :
```
from torch.nn.utils import prune
pruning_percentage = 0.40
parameters_to_prune=((model.encoder.0.mconv.0.conv,'weight'),(model.encoder.0.mconv.1.conv,'weight'))
prune.global_unstructured(
    parameters_to_prune,
    pruning_method=prune.L1Unstructured,
    amount=pruning_percentage, # Specifying the percentage
)
```
the error was : 
```
    parameters_to_prune=((model.encoder.0.mconv.0.conv,'weight'),(model.encoder.0.mconv.1.conv,'weight'))
                                       ^
   SyntaxError: invalid syntax
```

**Reply by titu1994** [reply]
To access the state, you have fo call s = model.state_dict() and specificity provide [s[...], s[...]] as parameters to prune. ... Must be string names to access the params of the state dict.

**Reply by amoghatr** [reply]
`prune.l1_unstructured(s['encoder.0.mconv.0.conv.weight'],name='weight',amount=0.3)`
```
AttributeError                            Traceback (most recent call last)
<ipython-input-25-0da9827c510a> in <module>
----> 1 prune.l1_unstructured(s['encoder.0.mconv.0.conv.weight'],name='weight',amount=0.3)

/opt/conda/lib/python3.8/site-packages/torch/nn/utils/prune.py in l1_unstructured(module, name, amount, importance_scores)
    921         odict_keys(['bias', 'weight_orig', 'weight_mask'])
    922     """
--> 923     L1Unstructured.apply(
    924         module, name, amount=amount, importance_scores=importance_scores
    925     )

/opt/conda/lib/python3.8/site-packages/torch/nn/utils/prune.py in apply(cls, module, name, amount, importance_scores)
    560                 If unspecified or None, the module parameter will be used in its place.
    561         """
--> 562         return super(L1Unstructured, cls).apply(
    563             module, name, amount=amount, importance_scores=importance_scores
    564         )

/opt/conda/lib/python3.8/site-packages/torch/nn/utils/prune.py in apply(cls, module, name, importance_scores, *args, **kwargs)
    145             return method
    146 
--> 147         method = _get_composite_method(cls, module, name, *args, **kwargs)
    148         # at this point we have no forward_pre_hooks but we could have an
    149         # active reparametrization of the tensor if another pruning method

/opt/conda/lib/python3.8/site-packages/torch/nn/utils/prune.py in _get_composite_method(cls, module, name, *args, **kwargs)
    104             # assert this using `found`
    105             hooks_to_remove = []
--> 106             for k, hook in module._forward_pre_hooks.items():
    107                 # if it exists, take existing thing, remove hook, then
    108                 # go through normal thing

AttributeError: 'Tensor' object has no attribute '_forward_pre_hooks'
```

**Reply by amoghatr** [reply]
I don't think this is the correct approach, since prune expects an nn.Module argument

**Reply by titu1994** [reply]
Try to follow the tutorials- they pass full module - https://pytorch.org/tutorials/intermediate/pruning_tutorial.html

**Reply by titu1994** [reply]
Oh right you got the same answer. So try with the full module - model.encoder.

**Reply by amoghatr** [reply]
Hi, could you link me to any documentation on quantization? I tried /examples/asr/quantization/speech_to_text_calibrate.py but that did not reduce the model size at all.

**Reply by titu1994** [reply]
I don't know much about pruning to be of help. We don't use it.
